[SPARK-10641][SQL] Add Skewness and Kurtosis Support

sethah opened the pull request:
Implementing skewness and kurtosis support based on following algorithm:
https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Higher-order_statistics

sethah committed:
Added skewness and kurtosis aggregate functions

sethah committed:
Adding kahan updates to higher order aggregate stats

sethah committed:
adding zero division protection

sethah committed:
Adding order check to reduce calculation overhead

sethah committed:
style and scaladoc fixes

sethah committed:
updating kurtosis test

AmplabJenkins commented:
Can one of the admins verify this patch?

sethah commented:
A few notes:
* I wrote this implementation before stddev was merged, and hence I ended up with a slightly different implementation. The differences are mostly syntactical (the algorithms for computing second moment are the same), except for how the subclasses are implemented. Once we figure out the best way to implement the subclasses, I will merge the implementations to remove the overlap
* (https://issues.apache.org/jira/browse/SPARK-10953)[SPARK-10953] Calls for benchmarking codegen vs scala implementations of univariate stats. The results of that may affect this PR.
* I did not implement the skewness and kurtosis functions using the old way with AggregateFunction1, since I think that will be removed. I can implement this separately if needed.

sethah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -88,6 +88,276 @@ case class Average(child: Expression) extends AlgebraicAggregate {
+  }
+}
+
+//case class Average(child: Expression) extends CentralMomentAgg(child) {
```
Aggregate functions which depend on lower order moments can easily be implemented using the CentralMomentAgg base class. I have commented them out for now, but left them for discussion purposes.

sethah commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala:
```diff
@@ -221,4 +221,40 @@ class DataFrameAggregateSuite extends QueryTest with SharedSQLContext {
+      testData2.agg(skewness('a)),
+      Row(0.0))
+
+    checkAnswer(
```
This test needs a tolerance added in order for it to pass. Appreciate any tips on how to do this.

JihongMA replied:
@sethah currently checkAnswer only supports exact match, looking at the code we can expand it to accept an error bound in QueryTest.scala.

holdenk replied:
So there are a few options for how to go about this. In spark-testing-base (which I'd like to eventually merge some of the functionality of) I made a modified version of checkAnswer which also takes tolerance and does tolerance based matching for floating point data (see https://github.com/holdenk/spark-testing-base/blob/master/src/main/scala/com/holdenkarau/spark/testing/DataFrameSuiteBase.scala & approxEqualDataFrames + approxEquals for comparing the rows). The other option would be, since for these tests its just a single value, just collect the result back, extract the value and do a regular assertion with tolerance. 

I think collecting the result and just checking the single value is probably the best plan for this PR and eventually maybe looking at a more general version of check answer.

sethah replied:
@holdenk @JihongMA Thanks for the suggestions! I updated `SQLQuerySuite` with a private `checkAnswerWithTol` helper function that solves the problem for the specific tests written so far. However, there are tests in `DataFrameAggregateSuite` that could fail if the test data ever changes - they too could benefit from a `checkAnswer` function that accepts a tolerance. I'm not sure if we can leave as is for now or perhaps the helper function should be added to `QueryTest` object so that it can be used in both test suites. Open to suggestions...

holdenk replied:
I think sharing `checkAnswerWithTol` as a helper function in `QueryTest` sounds like a good plan, I'm not super sure on the name (since checkAnswer works with an expected Row and the signature is a bit further off than just adding a tolerance param), but it seems probably ok with the current name.

sethah commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -734,10 +750,30 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
+    )
+  }
+
+  test("kurtosis") {
```
Ditto about tolerancing.

sethah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -88,6 +88,276 @@ case class Average(child: Expression) extends AlgebraicAggregate {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM3 = if (maxMoment >= 3) {
```
maxMoment variable is implemented with intent to avoid extra computation. For instance, if variance function calls `CentralMomentAgg` then the 3rd and 4th moments do not need to be computed. Would appreciate feedback/criticism of this implementation.

JihongMA replied:
@sethah would define Enumeration to represent the moment better?

sethah committed:
converting from codegen to imperative aggregate

mengxr commented:
@sethah, are you working on this? My suggestion is to implement skewness and kurtosis in a single aggregate function that implements ImperativeAggregate based on the discussion in SPARK-10953, but do not change existing implementation of standard deviation and variance. This makes the PR simpler and easier to review and merge.

sethah committed:
cleaning up and style fixes

sethah commented:
@mengxr I am working on it, and I have incorporated changes from the note you posted on the [Jira](https://issues.apache.org/jira/browse/SPARK-10641) - thanks for that! I pushed some changes. I had already changed the existing implementation of stddev unfortunately, but I can revert those changes so that the PR is a bit smaller. I was thinking that since variance hasn't been merged yet, that I could include it in the PR, but if you think it's best to leave it out I can do that as well.

I tested the imperative versions vs the codegen versions I had before and they are significantly faster. I will work on getting the changes for stddev reverted so that it will be ready for review.

JihongMA commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,304 @@ object HyperLogLogPlusPlus {
+      val updateValue = v match {
+        case d: java.lang.Number => d.doubleValue()
+        case _ => 0.0
+      }
```
please double check if it will hanle all numeric types correctly, e.g: Decimal?

JihongMA commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,304 @@ object HyperLogLogPlusPlus {
+
+  override def aggBufferSchema: StructType = StructType.fromAttributes(aggBufferAttributes)
+
+  def cloneBufferAttributes: Seq[Attribute] = aggBufferAttributes.map(_.newInstance())
```
cloneBufferAttributes is no longer needed.

sethah committed:
cast child expression to double type

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
   // scalastyle:on
 }
+
+abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable {
```
Missing API doc. It is useful to cite the Wikipedia page. My version can be find here: http://arxiv.org/abs/1510.04923.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+        case d: Double => d
+        case _ => 0.0
+      }
+      val currentM0 = buffer.getDouble(mutableAggBufferOffset)
```
It is easier to call this `n` and make it a `var`.

sethah replied:
Done.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+        case _ => 0.0
+      }
+      val currentM0 = buffer.getDouble(mutableAggBufferOffset)
+      val currentM1 = buffer.getDouble(mutableAggBufferOffset + 1)
```
* This is not exactly the first central moment, which is always zero. So better to call it `mean`.
* Make it a `var` and same applies to `currentM2`, `currentM3`, and `currentM4`.
* Since the updates are in place, so there is no need to call it `currentM2`. Just `m2` should be sufficient.

sethah replied:
Just to be clear, if we make `m2`, `m3`, ... as vars, we'll need to initialize them outside the if statements like

```scala
var m2 = 0.0
var m3 = 0.0
var m4 = 0.0
```

Then we can do the `buffer.getDouble` and `buffer.setDouble` inside the if statement to avoid doing these when possible.

mengxr replied:
Yes. This should simplify the implementation, and reduce the overhead of computing lower-order statistics to a few if checks.

sethah replied:
Changed variable names and declared as vars.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+      val currentM0 = buffer.getDouble(mutableAggBufferOffset)
+      val currentM1 = buffer.getDouble(mutableAggBufferOffset + 1)
+      val currentM2 = buffer.getDouble(mutableAggBufferOffset + 2)
+      val currentM3 = buffer.getDouble(mutableAggBufferOffset + 3)
```
not needed if `maxMoment >= 3`

sethah replied:
moved unnecessary gets and sets into if statements

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+      val currentM4 = buffer.getDouble(mutableAggBufferOffset + 4)
+
+      val updateM0 = currentM0 + 1.0
+      val delta = updateValue - currentM1
```
`delta`, `deltaN`, `delta2`, and `deltaN` are frequently used. We can store them as a `private[this] var`, which should simplify the logic a little bit.

sethah replied:
done.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+      val delta = updateValue - currentM1
+      val deltaN = delta / updateM0
+
+      val updateM1 = currentM1 + delta / updateM0
```
This is the same as after we rename properly and change `val` to `var`:

~~~
mean += deltaN
~~~

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+      val updateM1 = currentM1 + delta / updateM0
+      val updateM2 = if (maxMoment >= 2) {
+        currentM2 + delta * (delta - deltaN)
+      } else {
```
We don't need the `else` branch because it is not used if `maxMoment >= 2`.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+      } else {
+        0.0
+      }
+      val delta2 = delta * delta
```
Those computation are not needed if `maxMoment < 3`. So we should move then into the `if (maxMoment >= 3) {` block. Again, changing them to private vars helps here.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+      buffer.setDouble(mutableAggBufferOffset, updateM0)
+      buffer.setDouble(mutableAggBufferOffset + 1, updateM1)
+      buffer.setDouble(mutableAggBufferOffset + 2, updateM2)
+      buffer.setDouble(mutableAggBufferOffset + 3, updateM3)
```
not needed if `maxMoment < 3`.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+
+  /** Merge two central moment buffers. */
+  override def merge(buffer1: MutableRow, buffer2: InternalRow): Unit = {
+    val zeroMoment1 = buffer1.getDouble(mutableAggBufferOffset)
```
`n1`

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+  override def merge(buffer1: MutableRow, buffer2: InternalRow): Unit = {
+    val zeroMoment1 = buffer1.getDouble(mutableAggBufferOffset)
+    val zeroMoment2 = buffer2.getDouble(inputAggBufferOffset)
+    val firstMoment1 = buffer1.getDouble(mutableAggBufferOffset + 1)
```
`mean1`

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+
+    val firstMoment = firstMoment1 + deltaN * zeroMoment2
+
+    val secondMoment = if (maxMoment >= 2) {
```
It is useful to provide a link to the wikipedia page.

sethah replied:
Do you think we should link to page in the code as well as in the API doc?

mengxr replied:
I think it is useful to put a link inline because that wikipedia page is long.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+
+    val secondMoment = if (maxMoment >= 2) {
+      secondMoment1 + secondMoment2 + delta * deltaN * zeroMoment1 * zeroMoment2
+    } else {
```
not necessary to have the `else` branch

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
+    val M2 = buffer.getDouble(mutableAggBufferOffset + 2)
+
+    if (M0 == 0.0) {
+      0.0
```
Should be `NaN` instead of `0.0` when `n = 0` or `n = 1`. Check R's output.

sethah replied:
R and SciPy differ on this. R outputs NaN in divide by zero situations where SciPy outputs 0. I will assume we stick with R implementation.

mengxr replied:
R is more correct on statistics, I assume:)

mengxr commented:
@sethah Besides inline comments, we can also think about how to simplify the class inheritance. It would be better if we don't need to touch `InternalRow` in subclasses. Also, `Variance` is the same as `VarianceSamp`, shall we make one inherit another with a different `prettyName`?

mengxr commented:
add to whitelist

sethah commented:
@mengxr I think having `VarianceSamp` inherit from `Variance` will be a fine solution. I'm not clear on why it is better not to touch `InternalRow` in the subclasses, but if we must avoid it, one solution could be

```scala
def eval(buffer: InternalRow): Any = this match {
      case _: VariancePop => { }
      case _: Variance => { }
      case _: Skewness => { }
      case _: Kurtosis => { }
```

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #43961 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43961/consoleFull)** for PR 9003 at commit [`b180a28`](https://github.com/apache/spark/commit/b180a2857716ccf0e7c32d5aca0417f5a515d2cb).

SparkQA commented:
**[Test build #43961 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43961/consoleFull)** for PR 9003 at commit [`b180a28`](https://github.com/apache/spark/commit/b180a2857716ccf0e7c32d5aca0417f5a515d2cb).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `case class Stddev(child: Expression) extends CentralMomentAgg(child) `\n  * `case class StddevSamp(child: Expression) extends CentralMomentAgg(child) `\n  * `case class StddevPop(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Variance(child: Expression) extends CentralMomentAgg(child) `\n  * `case class VarianceSamp(child: Expression) extends CentralMomentAgg(child) `\n  * `case class VariancePop(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Skewness(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Kurtosis(child: Expression) extends CentralMomentAgg(child) `\n  * `abstract class CentralMomentAgg1(child: Expression)`\n  * `case class Kurtosis(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class Skewness(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class Variance(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class VariancePop(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class VarianceSamp(child: Expression) extends CentralMomentAgg1(child) `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43961/
Test FAILed.

sethah committed:
using vars for aggregator

mengxr commented:
@sethah The storage of `InternalRow` is an implicit contract. Exposing `InternalRow` to subclasses makes the code harder to understand. For example, to confirm the following code, one needs to go back to the implementation of `CentralMomentAgg`.

~~~scala
val M0 = buffer.getDouble(mutableAggBufferOffset)
val M2 = buffer.getDouble(mutableAggBufferOffset + 2)
val M4 = buffer.getDouble(mutableAggBufferOffset + 4)
~~~

I suggest adding an abstract method to `CentralMomemtAgg` called `Double getStatistic(n: Double, mean: Double, moments: Array[Double])` and implement `eval` in `CentralMomentAgg`:

~~~scala
override final def eval(buffer: InternalRow): Any = {
  val n = buffer.getDouble(offset)
  val mean = buffer.getDouble(offset + 1)
  val moments = Array.ofDim(maxMomemt)
  moments[0] = 1.0
  moments[1] = 0.0
  // fill in other moments up to maxMoment
  getStatistic(n, mean, moments)
}
~~~

Then we can hide the storage of `InternalRow` from subclasses.

sethah committed:
restructuring eval method

sethah committed:
style cleanup

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44039 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44039/consoleFull)** for PR 9003 at commit [`4ae234e`](https://github.com/apache/spark/commit/4ae234e3c8385e3149937821f5cdcee8fb85c053).

SparkQA commented:
**[Test build #44039 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44039/consoleFull)** for PR 9003 at commit [`4ae234e`](https://github.com/apache/spark/commit/4ae234e3c8385e3149937821f5cdcee8fb85c053).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44039/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

sethah committed:
reverting some stddev changes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44046 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44046/consoleFull)** for PR 9003 at commit [`62c515c`](https://github.com/apache/spark/commit/62c515cad1e2fb712256601f139d098fa4fb3749).

sethah commented:
@mengxr Addressed all comments and rearranged the class inheritances for case classes. I made another abstract class `SecondMoment` which all `Variance` classes inherit from - I think this will make the addition of the `StdDev` classes very simple.

Items still to be addressed: 

* I have not implemented using the `AggregateFunction1` interface since I believe it will be removed. Right now, I have added placeholders that just copy the code for `Average` so that the current code will compile. I'm not sure if I need to implement this or if we can just wait until the `AggregateFunction1` interface is removed.
* I need to add tolerancing to the `checkAnswer` function in SQL test suite because aggregate stats do not result in exact answers for the test data (e.g. 0.0 != 5e-17). I will work on this soon.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44045/
Test FAILed.

sethah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -857,3 +857,329 @@ object HyperLogLogPlusPlus {
+
+  override final def eval(buffer: InternalRow): Any = {
+    val n = buffer.getDouble(mutableAggBufferOffset)
+    val moments = sufficientMoments.map { momentIdx =>
```
Using a `sufficientMoments` variable defined in subclasses we can make sure we don't get values from the buffer that are unused for a particular aggregate. This results in defining an Array of sufficient moments in the subclasses like

```scala
override protected val sufficientMoments = Array(2, 4) // second and fourth moments needed
```

Would appreciate feedback on this approach.

JihongMA replied:
noticed you override getStatistics() in all sub-class, thought this adds complexity of reading this code, simply override eval() in specific sub-class (secondmoment, skewness, kurtosis) would be clear and starightforward.

sethah replied:
Per Xiangrui's comments about passing buffer to subclasses, I think it's necessary to do it this way. We don't have to pass the InternalRow object to the subclasses, and just pass the sufficient moments instead. See below discussion for more details.

sethah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -857,3 +857,329 @@ object HyperLogLogPlusPlus {
+  }
+}
+
+abstract class SecondMoment(child: Expression) extends CentralMomentAgg(child) {
```
All `StdDev` and `Variance` subclasses can extend `SecondMoment` abstract class since their `getStatistic` methods are nearly identical.

mengxr replied:
This class actually complicates the implementation. Most of the subclasses only few lines of code:

~~~scala
case class VarianceSamp(child: Expression) extends CentralMomentAgg(child) {
  override def momentOrder = 2
  override def getStatistics(n: Double, mean: Double, centralMoments: Array[Double]): Double = {
    if (n == 0 || n == 1) Double.NaN else centralMoments(2) / (n - 1.0)
  }
}
~~~

SparkQA commented:
**[Test build #44046 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44046/consoleFull)** for PR 9003 at commit [`62c515c`](https://github.com/apache/spark/commit/62c515cad1e2fb712256601f139d098fa4fb3749).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `abstract class SecondMoment(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Variance(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VarianceSamp(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VariancePop(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Skewness(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `abstract class CentralMomentAgg1(child: Expression)`\n  * `case class Kurtosis(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class Skewness(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class Variance(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class VariancePop(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class VarianceSamp(child: Expression) extends CentralMomentAgg1(child) `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44046/
Test FAILed.

JihongMA commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -857,3 +857,329 @@ object HyperLogLogPlusPlus {
+    while (aggIndex < numMoments) {
+      buffer.setDouble(mutableAggBufferOffset + aggIndex, 0.0)
+      aggIndex += 1
+    }
```
for (aggIndex <- 0 until numMoments) {
  buffer.setDouble(mutableAggBufferOffset + aggIndex, 0.0)
}

sethah replied:
Updated.

JihongMA commented:
Seth, no need to implement the old interface, simply put a placeholder for resolving it is sufficient, which will go away when this code path is removed. 

here is an example

case class Range(child: Expression) extends UnaryExpression with AggregateExpression {
+  override def nullable: Boolean = false
+  override def dataType: DoubleType.type = DoubleType
+  override def foldable: Boolean = false
+  override def toString: String = s"RANGE($child)"
+}

sethah committed:
adding helper function for tests with tolerances

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
**[Test build #44119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44119/consoleFull)** for PR 9003 at commit [`1594aba`](https://github.com/apache/spark/commit/1594abafec18dd0ba1c67c7a68c8ff017f88660a).

SparkQA commented:
**[Test build #44119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44119/consoleFull)** for PR 9003 at commit [`1594aba`](https://github.com/apache/spark/commit/1594abafec18dd0ba1c67c7a68c8ff017f88660a).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `abstract class SecondMoment(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Variance(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VarianceSamp(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VariancePop(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Skewness(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `abstract class CentralMomentAgg1(child: Expression)`\n  * `case class Kurtosis(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class Skewness(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class Variance(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class VariancePop(child: Expression) extends CentralMomentAgg1(child) `\n  * `case class VarianceSamp(child: Expression) extends CentralMomentAgg1(child) `\n

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44119/
Test FAILed.

sethah committed:
more generic tests with tolerance function and placeholders for AggregateFunction1

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

sethah commented on sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala:
```diff
@@ -168,6 +191,28 @@ object QueryTest {
+   * @param expectedAnswer the expected result in a[[Row]].
+   * @param absTol the absolute tolerance between actual and expected answers.
+   */
+  protected def checkAggregatesWithTol(actualAnswer: Row, expectedAnswer: Row, absTol: Double) = {
```
@holdenk Changed name to `checkAggregatesWithTol` and moved to query test. We can now use this for the various aggregate testing. I think ultimately a more elegant solution is needed, but this provides a way to add tolerancing for now.

SparkQA commented:
**[Test build #44164 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44164/consoleFull)** for PR 9003 at commit [`2c9046f`](https://github.com/apache/spark/commit/2c9046ff952d80ff6c9382fc9e1ba31dc8ade7b8).

SparkQA commented:
**[Test build #44164 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44164/consoleFull)** for PR 9003 at commit [`2c9046f`](https://github.com/apache/spark/commit/2c9046ff952d80ff6c9382fc9e1ba31dc8ade7b8).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `abstract class SecondMoment(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Variance(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VarianceSamp(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VariancePop(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Skewness(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Skewness(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Variance(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VariancePop(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VarianceSamp(child: Expression) extends UnaryExpression with AggregateExpression `\n

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44164/
Test FAILed.

sethah committed:
style and readability updates

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
**[Test build #44165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44165/consoleFull)** for PR 9003 at commit [`e0c5c88`](https://github.com/apache/spark/commit/e0c5c8864e2c75d612f2723020850894de8f9db5).

SparkQA commented:
**[Test build #44165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44165/consoleFull)** for PR 9003 at commit [`e0c5c88`](https://github.com/apache/spark/commit/e0c5c8864e2c75d612f2723020850894de8f9db5).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `abstract class SecondMoment(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Variance(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VarianceSamp(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VariancePop(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Skewness(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Skewness(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Variance(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VariancePop(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VarianceSamp(child: Expression) extends UnaryExpression with AggregateExpression `\n

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44165/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44179/consoleFull)** for PR 9003 at commit [`3ef2faa`](https://github.com/apache/spark/commit/3ef2faaa83104273ab185a3190f2d133a04693bb).

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+/**
+ * A central moment is the expected value of a specified power of the deviation of a random
+ * variable from the mean. Central moments are often used to characterize the properties of about
+ * the shape of a distribution.
```
Please also document the behavior for `null` and `NaN` values.

mengxr replied:
This is not for the case when `n = 0` or `n =1` but when we have `null` or `NaN` in the values. Are we ignoring them or outputting `NaN` directly?

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+  /**
+   * Array of sufficient moments need to compute the aggregate statistic.
+   */
+  protected def sufficientMoments: Array[Int]
```
This is not necessary because the k-th order moment update depends on all lower order moments, even we don't use it in the final formulation. Since `eval` is only called once, simplicity and readability is more important than performance. So having an `Array` with all central moments is better here.

sethah replied:
Removed this def and instead pass all moments up to the maximum moment to the `getStatistic` function.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+  /**
+   * The number of central moments to store in the buffer.
+   */
+  private[this] val numMoments = 5
```
`numMoments` -> `bufferSize`

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+  }
+
+  // frequently used values for online updates
+  private[this] var delta = 0.0
```
Shall we move all variables before methods?

sethah replied:
done.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+    if (v != null) {
+      val updateValue = v match {
+        case d: Double => d
+        case _ => 0.0
```
Could it happen? If this is not expected, we should throw an exception instead of treating it as `0.0`.

sethah replied:
Looking at the code, `Cast.eval` should return the correct type or null, so this extra case statements has been removed.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+      }
+      var n = buffer.getDouble(mutableAggBufferOffset)
+      var mean = buffer.getDouble(mutableAggBufferOffset + 1)
+      var m2 = 0.0
```
Those vars could be `private[this]` member vars.

sethah replied:
Added as `private[this]` vars

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+      deltaN = delta / n
+      mean += deltaN
+      buffer.setDouble(mutableAggBufferOffset, n)
+      buffer.setDouble(mutableAggBufferOffset + 1, mean)
```
If the `mutableAggBufferOffset` is set at the constructor, can we define `private[this]` constant for them, e.g., 'nOffset', 'meanOffset`? This saves some computations.

sethah replied:
@mengxr If we do this, I'm wondering if it doesn't make sense to store constants for all moments offset. It would help get rid of ambiguous integer offsets and should be fine if we don't plan to extend this to arbitrary order moments.

```scala
private[this] val nOffset = mutableAggBufferOffset
...
private[this] val kurtosisOffset = mutableAggBufferOffset + 4
```

mengxr replied:
I don't think we are going to support arbitrary-order moments. Kurtosis should be sufficient:)

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+      delta = updateValue - mean
+      deltaN = delta / n
+      mean += deltaN
+      buffer.setDouble(mutableAggBufferOffset, n)
```
minor: move this line right after `n += 1.0`?

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+  /**
+   * Compute aggregate statistic from sufficient moments.
+   */
+  def getStatistic(n: Double, moments: Array[Double]): Double
```
* Again, `n`, `mean`, and `centralMoments: Array[Double]` is better here. This is only called once, performance is not as important as readability and simplicity. See my previous comments on the API.
* Need API doc for `moments`.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,327 @@ object HyperLogLogPlusPlus {
+  }
+}
+
+case class Kurtosis(child: Expression, mutableAggBufferOffset: Int = 0,
```
* Need to document which Kurtosis we implemented.
* Chop down args to match Spark code style.

sethah replied:
Done.

SparkQA commented:
**[Test build #44179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44179/consoleFull)** for PR 9003 at commit [`3ef2faa`](https://github.com/apache/spark/commit/3ef2faaa83104273ab185a3190f2d133a04693bb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `abstract class SecondMoment(child: Expression) extends CentralMomentAgg(child) `\n  * `case class Variance(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VarianceSamp(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class VariancePop(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Skewness(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression, mutableAggBufferOffset: Int = 0,`\n  * `case class Kurtosis(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Skewness(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Variance(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VariancePop(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VarianceSamp(child: Expression) extends UnaryExpression with AggregateExpression `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44179/
Test FAILed.

sethah committed:
addressing feedback

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44234/consoleFull)** for PR 9003 at commit [`fd3f4d6`](https://github.com/apache/spark/commit/fd3f4d6f9ba5124406a7078c9e7991bf91abdad6).

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,330 @@ object HyperLogLogPlusPlus {
+  /**
+   * Compute aggregate statistic from sufficient moments.
+   * @param centralMoments Length `momentOrder + 1` array of central moments needed to
+   *                       compute the aggregate stat.
```
It is useful to mention that whether they are divided by `n` or not.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,330 @@ object HyperLogLogPlusPlus {
+    val n = buffer.getDouble(nOffset)
+    val mean = buffer.getDouble(meanOffset)
+    val moments = Array.ofDim[Double](momentOrder + 1)
+    moments(0) = n
```
Thought `moments(0)` and `moments(1)` are just placeholders, we should use the correct values `moments(0) = 1.0` and `moments(1) = 0.0`.

sethah replied:
Corrected.

mengxr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,330 @@ object HyperLogLogPlusPlus {
+    require(moments.length == momentOrder + 1,
+      s"$prettyName requires ${momentOrder + 1} central moment, received: ${moments.length}")
+
+    if (n == 0.0 || n == 1.0) Double.NaN else moments(2) / n
```
Remove `|| n == 1.0`. Population variance should be `0.0` when `n == 1.0`, which is the same as `moments(2) / n`.

sethah replied:
Updated

mengxr commented:
LGTM except minor inline comments. Ping @yhuai for a final pass.

SparkQA commented:
**[Test build #44234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44234/consoleFull)** for PR 9003 at commit [`fd3f4d6`](https://github.com/apache/spark/commit/fd3f4d6f9ba5124406a7078c9e7991bf91abdad6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44234/
Test FAILed.

sethah committed:
correcting error in merge function

sethah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -930,3 +930,332 @@ object HyperLogLogPlusPlus {
+    n = n1 + n2
+    buffer1.setDouble(nOffset, n)
+    delta = mean2 - mean1
+    deltaN = if (n == 0.0) 0.0 else delta / n
```
Removed divide by zero case here, which was causing problems when number of partitions > number of samples.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44267/consoleFull)** for PR 9003 at commit [`cf8a14b`](https://github.com/apache/spark/commit/cf8a14bb24924079af2c30234a083a1e6c6d4c23).

sethah committed:
adding back some stddev codegen tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44272/consoleFull)** for PR 9003 at commit [`b86386a`](https://github.com/apache/spark/commit/b86386a78ab4b27ca9b6b84afe6db4973204e9e9).

SparkQA commented:
**[Test build #44267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44267/consoleFull)** for PR 9003 at commit [`cf8a14b`](https://github.com/apache/spark/commit/cf8a14bb24924079af2c30234a083a1e6c6d4c23).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `case class Variance(child: Expression,`\n  * `case class VarianceSamp(child: Expression,`\n  * `case class VariancePop(child: Expression,`\n  * `case class Skewness(child: Expression,`\n  * `case class Kurtosis(child: Expression,`\n  * `case class Kurtosis(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Skewness(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Variance(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VariancePop(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VarianceSamp(child: Expression) extends UnaryExpression with AggregateExpression `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44267/
Test FAILed.

SparkQA commented:
**[Test build #44272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44272/consoleFull)** for PR 9003 at commit [`b86386a`](https://github.com/apache/spark/commit/b86386a78ab4b27ca9b6b84afe6db4973204e9e9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44272/
Test FAILed.

mengxr commented:
@sethah It seems that `VARIANCE` in HIVE returns the population variance. It is the same in MySQL: https://dev.mysql.com/doc/refman/5.0/en/group-by-functions.html#function_variance. So let's keep the same behavior. Then it is a little awkward that `stddev` is sample variance in our implementation. HIVE doesn't have `stddev`, while MySQL uses population stddev for `stddev`. I think it is more important to be compatible with existing SQL engines. We can address the `stddev` in a follow-up PR.

sethah committed:
changing variance to default to population variance

sethah commented:
@mengxr Corrected `variance` to yield the population variance. Tests should pass now.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44366/consoleFull)** for PR 9003 at commit [`3045e3b`](https://github.com/apache/spark/commit/3045e3b1d82ac73d154dc4d2165e920c74bdc118).

SparkQA commented:
**[Test build #44366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44366/consoleFull)** for PR 9003 at commit [`3045e3b`](https://github.com/apache/spark/commit/3045e3b1d82ac73d154dc4d2165e920c74bdc118).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `case class Variance(child: Expression,`\n  * `case class VarianceSamp(child: Expression,`\n  * `case class VariancePop(child: Expression,`\n  * `case class Skewness(child: Expression,`\n  * `case class Kurtosis(child: Expression,`\n  * `case class Kurtosis(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Skewness(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Variance(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VariancePop(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VarianceSamp(child: Expression) extends UnaryExpression with AggregateExpression `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44366/
Test FAILed.

sethah commented:
I guess this is still failing HiveComparisonTest due to small error in the variance. 

[info]   !== HIVE - 1 row(s) ==                                            == CATALYST - 1 row(s) ==
[info]   !500	130091	260.182	0	498	142.92680950752384	20428.072876000006   500	130091	260.182	0	498	142.92680950752384	20428.07287599999 (HiveComparisonTest.scala:433)

abs(20428.072876000006 - 20428.07287599999) = 1.4552e-11

Not sure if we need to change the test?

mengxr commented:
Had an offline discussion with @yhuai. We can remove `fetch_aggregate` from whitelist. We already tested all the functions there in SQL queries. We can create a new JIRA to add it back and handle small numerical differences, but it should be orthogonal to this PR.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala:
```diff
@@ -991,3 +991,73 @@ case class StddevFunction(
+  override def prettyName: String = "variance_samp"
+
+  override def toString: String = s"VAR_SAMP($child)"
+}
```
What will be the error message if we call this function when `spark.sql.useAggregate2=false`? It will be good to provide a meaning error message.

sethah replied:
@yhuai This is the error I get when calling one of the new `skewness` on a dataframe when `spark.sql.useAggregate2=false`: 

```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1.0 (TID 2, localhost): org.apache.spark.sql.catalyst.errors.package$TreeNodeException: Binding attribute, tree:{colName}#2
```

I'm not exactly sure where we ought to throw the error. I'd appreciate any tips on how to do this.

yhuai replied:
Can you change `with AggregateExpression` to `with AggregateExpression1`? Then, in the `newInstance` method, we throw an `UnsupportedOperationException` to let users know that they need to set `spark.sql.useAggregate2` to `true`.

sethah replied:
@yhuai Thanks for the suggestion! I think that's working now if you want to review it :)

sethah committed:
removing fetch_aggregation from whitelist

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44595 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44595/consoleFull)** for PR 9003 at commit [`ff363cc`](https://github.com/apache/spark/commit/ff363cca57e2b1c2bb28e281d014d33b930fd603).

SparkQA commented:
**[Test build #44595 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44595/consoleFull)** for PR 9003 at commit [`ff363cc`](https://github.com/apache/spark/commit/ff363cca57e2b1c2bb28e281d014d33b930fd603).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `case class Variance(child: Expression,`\n  * `case class VarianceSamp(child: Expression,`\n  * `case class VariancePop(child: Expression,`\n  * `case class Skewness(child: Expression,`\n  * `case class Kurtosis(child: Expression,`\n  * `case class Kurtosis(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Skewness(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class Variance(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VariancePop(child: Expression) extends UnaryExpression with AggregateExpression `\n  * `case class VarianceSamp(child: Expression) extends UnaryExpression with AggregateExpression `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44595/
Test PASSed.

sethah committed:
Throw UnsupportedOperationException for AggregateExpression1

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44611 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44611/consoleFull)** for PR 9003 at commit [`f49ce5c`](https://github.com/apache/spark/commit/f49ce5c0d6594f3b5348eda75e9aefed10617066).

SparkQA commented:
**[Test build #44611 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44611/consoleFull)** for PR 9003 at commit [`f49ce5c`](https://github.com/apache/spark/commit/f49ce5c0d6594f3b5348eda75e9aefed10617066).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable `\n  * `case class Variance(child: Expression,`\n  * `case class VarianceSamp(child: Expression,`\n  * `case class VariancePop(child: Expression,`\n  * `case class Skewness(child: Expression,`\n  * `case class Kurtosis(child: Expression,`\n  * `case class Kurtosis(child: Expression) extends UnaryExpression with AggregateExpression1 `\n  * `case class Skewness(child: Expression) extends UnaryExpression with AggregateExpression1 `\n  * `case class Variance(child: Expression) extends UnaryExpression with AggregateExpression1 `\n  * `case class VariancePop(child: Expression) extends UnaryExpression with AggregateExpression1 `\n  * `case class VarianceSamp(child: Expression) extends UnaryExpression with AggregateExpression1 `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44611/
Test PASSed.

mengxr commented:
Merge into master. Thanks! @JihongMA @sethah Could you send a PR to replace `Stddev` implementation using `CentralMomentAgg`? I ran the following and the difference is significant:

~~~scala
val df = sqlContext.range(100000000)
df.select(var_samp("id")).show();
df.select(stddev_samp("id")).show()
~~~

JihongMA commented:
Thanks @mengxr , I will send a PR for Stddev.
